### ISSUE TITLE
metabase high cpu usage v0.33.4

### ISSUE THREAD

**Issue by IklamA** [metabase/metabase#11392]
**Describe the bug**
For example if a user want to download something in Metabase we can see on the machine where Metabase is set up that the CPU usage increase to a high value and Metabase hanging on and can't used anymore.

**Logs**
In the Logs we then usually can see issues like gc overhead limit exceeded and on the machine we can see that the cpu usage increases to a high value:
<img width="901" alt="image" src="https://user-images.githubusercontent.com/40380839/69628720-1e4ad100-104c-11ea-997e-ef9219c2a5f9.png">

Metabase Config Information:
- Your browser and the version: IE, Firefox,Chrome
- Your operating system:Ubuntu 16.04
- Your databases: H2
- Metabase version: v0.33.4
- Metabase hosting environment: jar file on Ubuntu 16.04
- Java Version:
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (build 1.8.0_222-8u222-b10-1ubuntu1~16.04.1-b10)
OpenJDK 64-Bit Server VM (build 25.222-b10, mixed mode)
- Metabase internal database: H2 default
- Start parameters of the Metabase.jar:
....
/usr/bin/java -Xms1000m -Xmx5000m -jar /ed4/metabase/metabase.jar 
(fo testing we assign high memory )

A java staccktrace report many of the messages like this:
"Attach Listener" #107 daemon prio=9 os_prio=0 tid=0x00007f4960006000 nid=0x417b waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
"clojure-agent-send-off-pool-20" #106 prio=5 os_prio=0 tid=0x00007f495c108800 nid=0x4158 waiting on condition [0x00007f493eef3000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000687dd14c8> (a java.util.concurrent.SynchronousQueue$TransferStack)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:460)
	at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
	at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1073)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Thanks in advance for your help.

Rgs

**Comment by arthurfurlan** [comment]
Same issue here with v0.33.6

**Comment by mazameli** [comment]
Any reliable reproduction steps? How large are the downloads in terms of rows and columns that users are attempting to download which cause these performance hits? If you're able to backup your app db and try upgrading to 0.34.1 to see if that resolves things, that would also be helpful.

**Comment by arthurfurlan** [comment]
We solved the problem by simply downgrading to the previous version (0.32). It's certainly an issue of 0.33 version (we didn't tested 0.34).

**Comment by nautilus28** [comment]
We are having same issue with v0.35.3

**Comment by flamber** [comment]
@nautilus28 
Please post "Diagnostic Info" from Admin > Troubleshooting.
The logs from when the high CPU usage is happening. Admin > Troubleshooting > Logs.
And your system specs.

**Comment by nautilus28** [comment]
`{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.6+10-post-Debian-1bpo91",
    "java.vendor": "Debian",
    "java.vendor.url": "https://tracker.debian.org/openjdk-11",
    "java.version": "11.0.6",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.6+10-post-Debian-1bpo91",
    "os.name": "Linux",
    "os.version": "4.9.0-7-amd64",
    "user.language": "en",
    "user.timezone": "UTC"
  },
  "metabase-info": {
    "databases": [
      "exasol",
      "postgres",
      "bigquery",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.16"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-04-24",
      "tag": "v0.35.3",
      "branch": "v0.35.3",
      "hash": "01ca8d1"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}`

Our system is 32 vCPUs and 200GB RAM.

Sorry I can't expose logs because we might have some sensitive data.

I can see that there is one particular process using too much CPU
  
<img width="1414" alt="Screenshot 2020-05-22 at 14 09 00" src="https://user-images.githubusercontent.com/4472857/82670843-cb865680-9c35-11ea-9bdf-652a649b2a4a.png">

**Comment by flamber** [comment]
@nautilus28 
Metabase is one process - with many threads.

Then redact the logs for sensitive information. Without knowing what Metabase is doing, when you have high CPU, makes it really difficult to figure out the root cause of this.

If this is happening while people are downloading XLSX, then 0.36.0 will bring memory and performance improvements #12521

Or perhaps it's queries to a specific database type, where the driver is doing something strange?

**Comment by nautilus28** [comment]
The core Metabase thread is high CPU. And it happens when noone exporting XLSX. We do have a custom built driver (EXASOL), and it is highly in used, so I believe it is related to it.

**Comment by nautilus28** [comment]
`{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.6+10-post-Debian-1bpo91",
    "java.vendor": "Debian",
    "java.vendor.url": "https://tracker.debian.org/openjdk-11",
    "java.version": "11.0.6",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.6+10-post-Debian-1bpo91",
    "os.name": "Linux",
    "os.version": "4.9.0-7-amd64",
    "user.language": "en",
    "user.timezone": "UTC"
  },
  "metabase-info": {
    "databases": [
      "exasol",
      "postgres",
      "bigquery",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.16"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-08-10",
      "tag": "v0.35.4",
      "branch": "v0.35.4",
      "hash": "42a2a2a"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}`

We have increased 64 vCPUs, but the Metabase instance are still experiencing 100% CPU usage

**Comment by flamber** [comment]
@nautilus28 
But the question is still if this is caused by some changes in your custom build, or the custom driver, or something general.
You probably need to run with a CPU profiler, so you can see where the resources are being used.
You might want to read this too: https://www.metabase.com/blog/scaling-metabase/index.html

**Comment by nautilus28** [comment]
@flamber would you recommend what tool I could use to profile CPU usage?

**Comment by flamber** [comment]
@nautilus28 Profiling is such a vast subject, and I'm definitely not an expert in that, but YourKit and Flight Recorder comes to mind.
There has been several fixes in 0.36 that lowers the amount of information requested on certain pages, which lowers the CPU usage etc.
Again, it should be fairly simple for you just to look at the log and compare that to the CPU usage - that should give you a very good idea where the problem is.

**Comment by nautilus28** [comment]
I have managed to generate some profiling graphs from our production Metabase instance.

The graphs show that there are 2000 threads from streaming group. While I am pretty we don't have that much active users that the time I generate the profiling graphs (7pm). I have a feeling that these threads are leaks.

<img width="1334" alt="Screenshot 2020-08-11 at 19 36 22" src="https://user-images.githubusercontent.com/4472857/89935525-35e1de00-dc0a-11ea-83d0-92f460ed4941.png">
<img width="1328" alt="Screenshot 2020-08-11 at 19 36 27" src="https://user-images.githubusercontent.com/4472857/89935532-38443800-dc0a-11ea-9bb2-4a2eefba64a7.png">
<img width="1316" alt="Screenshot 2020-08-11 at 19 36 36" src="https://user-images.githubusercontent.com/4472857/89935537-39756500-dc0a-11ea-98db-cd4f4971dd81.png">

**Comment by Telofy** [comment]
I have a very similar problem. I freshly installed `Metabase v0.39.0.1 (47bb5f2 release-x.39.x)` (that’s what it says), created a few queries, and promptly it got very slow. I think it started when I created the second query. Now I have seven queries and one dashboard. There’s one query that takes ~ 3 s when I paste it into PSQL and takes ~ 50–90 s in Metabase. During those 50–90 s, Metabase doesn’t seem to send it to the DB – when I put in a typo, the error message also takes about a minute to be displayed. But the CPU usage of the Java process is at 130–180%.

I’m referencing one query with a lot of rows in another query via CTR and the `{{#1}}` syntax. In the Postgres log it looks like the referenced query gets executed in isolation, then nothing happens for a long time, and then the actual query (with the referenced query correctly inlined) gets executed.

Is there maybe a way to prevent that first execution of the referenced query? I don’t see what it’s for, and maybe the (seemingly needless?) evaluation of its returned rows takes so long?

**Comment by Telofy** [comment]
I’ve found a workaround:

At the end of my referenced query I added:

  limit
      case
          when (select true from pg_stat_activity where query like lower('%YOU MAY RUN NOW%')) then 1e9
          else 0
      end

Then I added a comment to the referencing query (or queries): `-- you may run now`.

1e9 is just a sufficiently big number of my purposes; `all` somehow didn’t work. The `lower()` bit is necessary so the `like` test matches only the comment and not its own string pattern.

**Comment by raj-saxena** [comment]
@Telofy https://github.com/metabase/metabase/issues/16257 seems related

**Comment by flamber** [comment]
There has been a lot of changes since - exports are now streamed. Closing, so the issue doesn't get confused with other problems.
If you are still experiencing this, then please create a new issue with all the details requested and steps-to-reproduce.
